### ISSUE TITLE
Handle conditions where SquirrelMail's addressbooks contain pipes within the data

### ISSUE THREAD

**Issue by loweryaustin** [loweryaustin/cpanel-squirrelmail-exporter#2]
Some users may add additional columns manually by adding pipes to the info column.

We need to detect this and handle the situation accordingly.

**Comment by dchisolm** [comment]
PR for one solution @ https://github.com/loweryaustin/cpanel-squirrelmail-exporter/pull/15

**Comment by loweryaustin** [comment]
@dchisolm I really appreciate your effort here! Although, there are a couple of problems that prevent me from merging.

1. This project must not contain any dependancies that are not already on a default cPanel installation. This is because the export process must be a one step process without installlation instructions. Adding dependancies increases documentation efforts and support burden.
2. I saw that the proposed method would successfully separate out columns by the pipe separator, but it would not handle the situation in a way that would be compatible for import. For instance, if there were sub columns, the headers in the resulting csv would no longer match up with the new columns. In Horde this would result in a confusing import experience because the info name, email, and other columns could end up holding random information. In roundcube I believe the import will outright fail because it needs valid headers for each column. Even further, the way that squirrelmail stores columns with sub columns changes the data set which I'll show a bit further down.

I apologize, I should have had this enhancement better documented, as well as contributer guidelines in the README. I was very suprised to see that someone took an interest in taking a shot at helping out. Thanks again for trying.

I took a closer look at this today and found that the squirrelmail interface allows users to put pipes in every column except for the nickname column. This means that columns that have "sub columns" are surrounded in double quotes which changes the way we'd need to handle this. Here is a sample data set:
```
[root@srv00001 ~]# cat /home/parkspawnee/.sqmaildata/l.knope\@parks.pawnee.gov.abook 
l.knope@parks.pawnee.gov|Leslie|Knope|l.knope@parks.pawnee.gov|
Ron|Ron|Swanson|r.swanson@parks.pawnee.gov|just some notes and things
KNICKKNAME|Merp|Lerp|email@email.tld|"Just Some INfo Here | Some Additional Info | ANd More info"
Nickname|"Nick | NAme "|"Last | name"|"nick@name.tld | Another@email.tld"|"Some | addy |info"
nick@name.tld|Austin|Texas|nick@name.tld|Some info 
```
Also someting to note about this is that the Info column (as labeled in the SquirrelMail interface) is actually labeled as Notes where I define the headers in the following line:
```
	my $csvAddressBook = '"Nickname","First Name","Last Name","Email","Notes"'."\r\n";
```

I chose "Notes" for that column specifically because that is a valid entry in Roundcube. If I had named the column "Info" as it is in SquirrelMail, Roundcube imports would fail.

In other words, if any extra columns are added, they must be added in a way that matches valid existing fields in Roundcube so that imports function properly.

**Comment by loweryaustin** [comment]
Also after looking at this and finding that Squirrelmail actually specificlaly allows for pipes to be added through the interface, I have realized that this is no longer an enahancement but a bug.

**Comment by dchisolm** [comment]
Hi,

I'll have to double check, but am fairly certain Text::CSV is included with cPanel, but maybe you would have to use #!/usr/local/cpanel/3rdparty/bin/perl instead of system perl.

Squirrelmail uses PHP's fputcsv/fgetcsv http://php.net/manual/en/function.fgetcsv.php and in perl Text:CSV works identical for this purpose.

This method always has the correct number of columns for a valid abook file, so the header would remain valid too. The PHP method used in squirrelmail and Text::CSV's see to that similarly. Any data that would prevent the record from not being separated correctly is quoted or escaped by these methods.

The example address book below was parsed using this and was imported successfully into the latest 1.4 version of roundcube. Earlier versions don't support the international characters though.

bojangles@example.com|Bo||bojangles@example.com|
BigP|Big|P|bigp@nothing.com|"The biggest | Person | Ever"
Quoteman|Quote|Man|"quot""ed@email.com"|"Has ""Quotes"""
Elvis|elvis|presley|elvis@presley.com|
用户@例子.广告|Chinese, Unicode||用户@例子.广告|
अजय@डाटा.भारत   |Hindi, Unicode||अजय@डाटा.भारत A|
квіточка@пошта.укр|Ukrainian, Unicode|квіточка@пошта.укр|
θσερ@εχαμπλε.ψομ|Greek, Unicode||θσερ@εχαμπλε.ψομ|
Dörte@Sörensen.example.com|German, Unicode||Dörte@Sörensen.example.com|
коля@пример.рф|Russian, Unicode||коля@пример.рф|
`
`

**Comment by loweryaustin** [comment]
Wow! You're right, using /usr/local/cpanel/3rdparty/bin/perl does the trick.

Thanks so much, I learned a couple things from your pull request and you made this problem a cinche to take care of. Greatly apreciated.

When I tested this, I needed to use /usr/local/cpanel/3rdparty/bin/perl in the one liner so the README needs to be updated to reference the new perl binary.

Could you update the README as well in your PR? Otherwise, you could submit to the develop branch and I'll do the little bit of cleanup required to get it merged into master.

**Comment by dchisolm** [comment]
I added another commit to change to cPanel perl. Please test it out!

Glad to be able to help. AFAIK Squirrelmail is already [discontinued from 78](https://blog.cpanel.com/the-death-of-squirrelmail/). So the work you've done here will continue to be valuable as people make the transition.

**Comment by loweryaustin** [comment]
Hopefully this make the transition easier!

This bug has been resolved via #15
